lib/forwarder/handler_test.go

```diff
@@ -0,0 +1,97 @@
+package forwarder
+
+import (
+	"context"
+	"github.com/stretchr/testify/require"
+	"tcplb/lib/core"
+	"tcplb/lib/slog"
+	"testing"
+)
+
+func TestClientIDFromContext(t *testing.T) {
+	parentCtx := context.Background()
+	c := core.ClientID{Namespace: "handler-test", Key: "a"}
+	childCtx := NewContextWithClientID(parentCtx, c)
+	cPrime, ok := ClientIDFromContext(childCtx)
+	require.True(t, ok)
+	require.Equal(t, c, cPrime)
+}
+
+func TestClientIDFromContextMissing(t *testing.T) {
+	ctx := context.Background()
+	_, ok := ClientIDFromContext(ctx)
+	require.False(t, ok)
+}
+
+func TestUpstreamsFromContext(t *testing.T) {
+	a := core.Upstream{Network: "handler-test", Address: "a"}
+	b := core.Upstream{Network: "handler-test", Address: "b"}
+	c := core.Upstream{Network: "handler-test", Address: "c"}
+	upstreams := core.NewUpstreamSet(a, b, c)
+
+	parentCtx := context.Background()
+	childCtx := NewContextWithUpstreams(parentCtx, upstreams)
+	upstreamsPrime, ok := UpstreamsFromContext(childCtx)
+	require.True(t, ok)
+	require.Equal(t, upstreams, upstreamsPrime)
+}
+
+func TestUpstreamsFromContextMissing(t *testing.T) {
+	ctx := context.Background()
+	_, ok := UpstreamsFromContext(ctx)
+	require.False(t, ok)
+}
+
+func TestClientIDAndUpstreamsFromContext(t *testing.T) {
+	// test that one context key doesn't overwrite the other one...
+
+	clientID := core.ClientID{Namespace: "handler-test", Key: "a"}
+
+	a := core.Upstream{Network: "handler-test", Address: "a"}
+	b := core.Upstream{Network: "handler-test", Address: "b"}
+	c := core.Upstream{Network: "handler-test", Address: "c"}
+	upstreams := core.NewUpstreamSet(a, b, c)
+
+	parentCtx := context.Background()
+
+	childCtx := NewContextWithUpstreams(
+		NewContextWithClientID(parentCtx, clientID),
+		upstreams)
+
+	clientIDPrime, ok := ClientIDFromContext(childCtx)
+	require.True(t, ok)
+	require.Equal(t, clientID, clientIDPrime)
+
+	upstreamsPrime, ok := UpstreamsFromContext(childCtx)
+	require.True(t, ok)
+	require.Equal(t, upstreams, upstreamsPrime)
+}
+
+// alwaysPanicHandler always panics when asked to Handle.
+type alwaysPanicHandler struct {
+	PanicValue string
+}
+
+func (h alwaysPanicHandler) Handle(ctx context.Context, conn AuthenticatedConn) {
+	panic(h.PanicValue)
+}
+
+func TestRecovererHandlerLogsPanics(t *testing.T) {
+	logger := &slog.RecordingLogger{}
+
+	h := &RecovererHandler{
+		Logger: logger,
+		Inner:  alwaysPanicHandler{PanicValue: "oh no!"},
+	}
+
+	ctx := context.Background()
+	var conn AuthenticatedConn
+
+	h.Handle(ctx, conn)
+
+	require.Len(t, logger.Events, 1)
+	event := logger.Events[0]
+	require.Equal(t, "error", event.Level)
+	require.Equal(t, "RecovererHandler: Unexpected panic!", event.Msg)
+	require.Equal(t, "oh no!", event.Details)
```

Comment:
I don't have anything bad to say about the implementation here, but testing for particular message output (or count / ordering / etc) always makes me a bit uneasy. Arguably it might disincentivize adding new logs if sometimes doing so breaks tests.

Reply:
that's a good point.  i've worked on software projects that had swathes of overly-sensitive automated tests that measured "did anything change" rather than "are the requirements satisfied", so i can emphasise.

i've "relaxed" the test a bit -- it now only cares that it can find exactly one panic recovery log event somewhere in what was logged. it still cares about some exact things that get logged, but i've factored out the duplication with constants, so it reduces toil if anyone tweaks some of them.  not quite what you were asking but some steps in the right direction.

Reply:
I think that's a big improvement, it's no longer sensitive to ordering of log events and the shared constants should make it much less brittle.